File: src/api/routes/global/extensions/{id}/versions.ts
import { globalAPIRouter } from "@/api"
import { time } from "@rjweb/utils"
import { and, desc, eq, inArray, or, sql } from "drizzle-orm"

export = new globalAPIRouter.Path('/')
	.http('GET', '/', (http) => http
		.document({
			description: 'Get the install-base of a blueprint extensions versions',
			responses: {
				200: {
					description: 'Success',
					content: {
						'application/json': {
							schema: {
								type: 'object',
								additionalProperties: {
									type: 'number'
								}
							}
						}
					}
				}
			}, parameters: [
				{
					name: 'id',
					in: 'path',
					description: 'The ID or identifier of the extension',
					required: true,
					schema: {
						anyOf: [
							{ type: 'integer' },
							{ type: 'string' }
						]
					}
				}
			]
		})
		.onRequest(async(ctr) => {
			const id = ctr.params.get('id', '')
			if (!id) return ctr.status(ctr.$status.BAD_REQUEST).print({ errors: ['Invalid ID'] })

			const idInt = parseInt(id)

			const [ extension ] = await ctr["@"].cache.use(`extension::${id}`, () => ctr["@"].database.select(ctr["@"].database.fields.extension)
					.from(ctr["@"].database.schema.extensions)
					.innerJoin(ctr["@"].database.schema.authors, eq(ctr["@"].database.schema.extensions.authorId, ctr["@"].database.schema.authors.id))
					.where(and(
						eq(ctr["@"].database.schema.extensions.hidden, false),
						or(
							!isNaN(idInt) ? eq(ctr["@"].database.schema.extensions.id, idInt) : undefined,
							eq(ctr["@"].database.schema.extensions.identifier, id)
						)
					)),
				time(5).m()
			)

			if (!extension) return ctr.status(ctr.$status.NOT_FOUND).print({ errors: ['Extension not found'] })

			const versionStats = await ctr["@"].cache.use(`versions::${extension.id}`, () => ctr["@"].database.select({
					version: sql<string>`ext->>'version'`.as('version'),
					percentage: sql<number>`(COUNT(*) * 100.0 / SUM(COUNT(*)) OVER ())::numeric(5,2)`.mapWith(Number).as('percentage')
				})
					.from(
						ctr["@"].database.select({
							ext: sql`jsonb_array_elements(data->'blueprint'->'extensions')`.as('ext')
						})
						.from(ctr["@"].database.schema.telemetryData)
						.where(and(
							inArray(
								ctr["@"].database.schema.telemetryData.id,
								ctr["@"].database.select({ id: ctr["@"].database.schema.telemetryPanelsWithLatest.latest.latestTelemetryDataId })
									.from(ctr["@"].database.schema.telemetryPanelsWithLatest)
							),
							sql`created > NOW() - INTERVAL '2 days'`
						))
						.as('subq')
					)
					.where(sql`ext->>'identifier' = ${extension.identifier}`)
					.groupBy(sql`ext->>'version'`)
					.orderBy(desc(sql`percentage`)),
				time(5).m()
			)

			return ctr.print(Object.fromEntries(versionStats.map((stat) => [
				stat.version,
				stat.percentage
			])))
		})
	)

File: src/routes/extensions/_extension_/versions.rs
use super::State;
use utoipa_axum::{router::OpenApiRouter, routes};

mod index {
    use crate::{models::Extension, routes::ApiError, routes::GetState};
    use axum::{extract::Path, http::StatusCode};
    use indexmap::IndexMap;
    use sqlx::Row;

    #[utoipa::path(get, path = "/", responses(
        (status = OK, body = HashMap<String, f64>),
        (status = NOT_FOUND, body = inline(ApiError)),
    ), params(
        ("extension" = String, Path, description = "the extension identifier or id")
    ))]
    pub async fn route(
        state: GetState,
        Path(extension): Path<String>,
    ) -> (StatusCode, axum::Json<serde_json::Value>) {
        let extension = state
            .cache
            .cached(&format!("extensions::{}", extension), 300, || async {
                match extension.parse::<i32>() {
                    Ok(id) => {
                        if id < 1 {
                            None
                        } else {
                            Extension::by_id(&state.database, id).await
                        }
                    }
                    Err(_) => Extension::by_identifier(&state.database, &extension).await,
                }
            })
            .await;

        if extension.is_none() {
            return (
                StatusCode::NOT_FOUND,
                axum::Json(serde_json::to_value(ApiError::new(&["extension not found"])).unwrap()),
            );
        }

        let mut versions: IndexMap<String, f64> = IndexMap::new();

        let data = sqlx::query(
            r#"
            SELECT
                ext->>'version' AS version,
                (COUNT(*) * 100.0 / SUM(COUNT(*)) OVER ())::float8 AS percentage
            FROM
                (
                    SELECT jsonb_array_elements(data->'blueprint'->'extensions') AS ext
                    FROM telemetry_data
                    WHERE
                        id IN (
                            SELECT telemetry_data.id
                            FROM telemetry_panels_with_latest
                            WHERE latest_telemetry_data_id = (
                                SELECT latest_telemetry_data_id
                                FROM telemetry_panels_with_latest
                                ORDER BY created DESC
                                LIMIT 1
                            )
                        )
                        AND created > NOW() - INTERVAL '2 days'
                ) AS subq
            WHERE ext->>'identifier' = $1
            GROUP BY ext->>'version'
            ORDER BY percentage DESC
            "#,
        )
        .bind(&extension.unwrap().identifier)
        .fetch_all(state.database.read())
        .await
        .unwrap();

        for row in data {
            versions.insert(
                row.get("version"),
                (row.get::<f64, &str>("percentage") * 100.0).round() / 100.0,
            );
        }

        (
            StatusCode::OK,
            axum::Json(serde_json::to_value(&versions).unwrap()),
        )
    }
}

pub fn router(state: &State) -> OpenApiRouter<State> {
    OpenApiRouter::new()
        .routes(routes!(index::route))
        .with_state(state.clone())
}
